resources/type-system/flow-analysis.md

```diff
@@ -0,0 +1,713 @@
+uses it to derive `after(N)`, `null(N)`, `notNull(N)`, `true(N)`, and
+`false(N)`.
+
+If `N` is an expression, and the following rules specify the values to be
+assigned to `true(N)` and `false(N)`, but do not specify values for `null(N)`,
+`notNull(N)`, or `after(N)`, then they are by default assigned as follows:
+  - `null(N) = unreachable(after(N))`.
+  - `notNull(N) = join(true(N), false(N))`.
+  - `after(N) = notNull(N)`.
+
+If `N` is an expression, and the following rules specify the values to be
+assigned to `null(N)` and `notNull(N)`, but do not specify values for `true(N)`,
+`false(N)`, or `after(N)`, then they are by default assigned as follows:
+  - `true(N) = notNull(N)`.
+  - `false(N) = notNull(N)`.
+  - `after(N) = join(null(N), notNull(N))`.
+
+If `N` is an expression, and the following rules specify the value to be
+assigned to `after(N)`, but do not specify values for `true(N)`, `false(N)`,
+`null(N)`, or `notNull(N)`, then they are all assigned the same value as
+`after(N)`.
+
+
+- **Variable or getter**: If `N` is an expression of the form `x` 
+  where the type of `x` is `T` then:
+  - If `T <: Never` then:
+    - Let `null(N) = unreachable(before(N))`.
+    - Let `notNull(N) = unreachable(before(N))`.
+  - Otherwise if `T <: Null` then:
+    - Let `null(N) = before(N)`.
+    - Let `notNull(N) = unreachable(before(N))`.
+  - Otherwise if `T` is non-nullable then:
+    - Let `null(N) = before(N)`.
+    - Let `notNull(N) = unreachable(before(N))`.
+  - Otherwise:
+    - Let `null(N) = promote(x, Null, before(N))`
+    - Let `notNull(N) = promoteToNonNull(x, before(N))`
+
+- **True literal**: If `N` is the literal `true`, then:
+  - Let `true(N) = before(N)`.
+  - Let `false(N) = unreachable(before(N))`.
+
+- **False literal**: If `N` is the literal `false`, then:
+  - Let `true(N) = unreachable(before(N))`.
+  - Let `false(N) = before(N)`.
+
+- **null literal**: If `N` is the literal `null`, then:
+  - Let `null(N) = before(N)`.
+  - Let `notNull(N) = unreachable(before(N))`.
+
+- **other literal**: If `N` is some other literal than the above, then:
+  - Let `null(N) = unreachable(before(N))`.
+  - Let `notNull(N) = before(N)`.
+
+- **throw**: If `N` is a throw expression of the form `throw E1`, then:
+  - Let `before(E1) = before(N)`.
+  - Let `after(N) = unreachable(after(E1))`
+
+- **Local-variable assignment**: If `N` is an expression of the form `x = E1`
+  where `x` is a local variable, then:
+  - Let `before(E1) = before(N)`.
+  - Let `after(N) = assign(x, E1, after(E1))`.
+  - Let `true(N) = assign(x, E1, true(E1))`.
+  - Let `false(N) = assign(x, E1, false(E1))`.
+  - Let `null(N) = assign(x, E1, null(E1))`.
+  - Let `notNull(N) = assign(x, E1, notNull(E1))`.
+
+- **operator==** If `N` is an expression of the form `E1 == E2` then:
+  - Let `before(E1) = before(N)`
+  - Let `before(E2) = after(E1)`
+  - Let `true(N) = join(join(null(E1), null(E2)),
```

Comment:
"Join" doesn't do what you want here--it takes two FlowAnalysis states that represent divergent control flow paths and forms the FlowAnalysis state if they were to both branch to a common execution point.  That's not what's happening with `==`.  So, for example, in the expression `a == (b = 0)`, `null(E1)` hasn't seen the assignment to `b` yet, so `join(null(E1), null(E2))` produces a model in which `b` is not necessarily definitely assigned.

What's currently implemented is much simpler: if Ei is the literal expression `null`, and Ej is a read of a variable, then `false(N) = promoteToNonNull(Ej, after(Ej))`, otherwise `false(N) = after(E2)`.  In either case `true(N) = after(E2)`.

It's possible that we could achieve what you want with restrict, e.g. something like `true(N) = join(restrict(null(E1), null(E2), ...), restrict(notNull(E1), notNull(E2), ...))`.  But I'd have to think pretty hard about it, and I'm not really convinced it would buy us enough functionality to be worth it compared to what's already implemented.

Reply:
Ack.  I made a note of this, and we can iterate and follow up.